<a href="https://colab.research.google.com/github/alxmart/Projeto_Python_Data_Science/blob/master/Hora%2Bda%2Bpratica.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sobre o notebook

A "hora da prática" é uma atividade do curso de Python para Data Science da Alura que contém diversos desafios envolvendo os aprendizados de cada aula.

O notebook "hora da pratica" é um notebook construído com espaço para as células dos desafios propostos no curso, com ele você tem espaço para construir e executar suas soluções

## Aula 1 - Introdução ao Python

### Questão 1

Imprima a frase `Escola de Dados da Alura!`

In [1]:
print('Escola de Dados da Alura !\n')

Escola de Dados da Alura !



### Questão 2
Imprima na tela seu nome e seu sobrenome seguindo a estrutura abaixo:
```
Nome: [seu nome]
Sobrenome: [seu sobrenome]
```


In [2]:
print('Nome: Luizão')
print('Sobrenome: Mart\n')

Nome: Luizão
Sobrenome: Mart



### Questão 3
Imprima o seu primeiro nome letra a letra. Por exemplo, nome é Mirla, então eu obtenho a seguinte saída:
```
M
I
R
L
A
```


In [3]:
print('L')
print('U')
print('I')
print('Z\n')

nome = "Luizão"

for letra in nome:
  print(letra)


L
U
I
Z

L
u
i
z
ã
o


### Questão 4
Imprima o dia do seu nascimento em formato `dia mês ano`. Os valores de dia e ano não podem estar entre aspas. O formato deve estar como no exemplo, supondo uma data de aniversário dia 28 de fevereiro de 2003:
```
28 fevereiro 2003
```

In [4]:
print(31,'fevereiro', 2155,"\n")

31 fevereiro 2155 



### Questão 5
Imprima em um único print o atual ano que você está fazendo esse curso, o valor do ano deve ser um dado numérico. A saída do print deve ser a seguinte:
```
Ano atual: [ano]
```


In [5]:
print('Ano atual:',2025,"\n")

Ano atual: 2025 



## Aula 2 - Manipulando dados no Python

### Coleta e amostragem de dados

#### Questão 1
Crie um programa que solicite à pessoa usuária digitar seu nome, e imprima “Olá, [nome]!”.

In [6]:
nome_pessoa = input('Por favor,digite seu nome:')
print(f'Olá, {nome_pessoa} !')

Por favor,digite seu nome:Homer
Olá, Homer !


#### Questão 2
Crie um programa que solicite à pessoa usuária digitar seu nome e idade, e imprima “Olá, [nome], você tem [idade] anos.”.

In [7]:
nome_pessoa = input('Por favor,digite seu nome:')
idade_pessoa = input('Por favor,digite sua idade:')
print(f'Olá, {nome_pessoa} Você tem {idade_pessoa} anos !')

Por favor,digite seu nome:McClaud
Por favor,digite sua idade:451
Olá, McClaud Você tem 451 anos !


#### Questão 3
Crie um programa que solicite à pessoa usuária digitar seu nome, idade e altura em metros, e imprima “Olá, [nome], você tem [idade] anos e mede [altura] metros!”.

In [8]:
nome_pessoa = input('Por favor,digite seu nome:')
idade_pessoa = input('Por favor,digite sua idade:')
altura_pessoa = input('Por favor,digite sua altura:')

print(f'Olá, {nome_pessoa} Você tem {idade_pessoa} anos e mede {altura_pessoa} metros')

Por favor,digite seu nome:Modrum
Por favor,digite sua idade:987
Por favor,digite sua altura:2.10
Olá, Modrum Você tem 987 anos e mede 2.10 metros


### Calculadora com operadores

#### Questão 1

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a soma dos dois valores.

In [9]:
valor1 = float(input(' Digite o primeiro valor: '))
valor2 = float(input(' Digite o segundo valor: '))
print("Soma dos valores: ", valor1 + valor2 )

 Digite o primeiro valor: 12
 Digite o segundo valor: 4
Soma dos valores:  16.0


#### Questão 2
Crie um programa que solicite três valores numéricos à pessoa usuária e imprima a soma dos três valores.

In [10]:
valor1 = float(input(' Digite o primeiro valor: '))
valor2 = float(input(' Digite o segundo valor: '))
valor3 = float(input(' Digite o terceiro valor: '))
print("Soma dos valores: ", valor1 + valor2 + valor3 )

 Digite o primeiro valor: 3
 Digite o segundo valor: 5
 Digite o terceiro valor: 4
Soma dos valores:  12.0


#### Questão 3
Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a subtração do primeiro pelo o segundo valor.

In [11]:
valor1 = float(input(' Digite o primeiro valor: '))
valor2 = float(input(' Digite o segundo valor: '))
print("Subtração do primeiro pelo segundo: ", valor1 - valor2 )

 Digite o primeiro valor: 10
 Digite o segundo valor: 7
Subtração do primeiro pelo segundo:  3.0


#### Questão 4

Crie um programa que solicite dois valores numéricos à pessoa usuária e imprima a multiplicação dos dois valores.


In [12]:
valor1 = float(input(' Digite o primeiro valor: '))
valor2 = float(input(' Digite o segundo valor: '))
print("Multiplicacao dos valores: ", valor1 * valor2 )

 Digite o primeiro valor: 5
 Digite o segundo valor: 2
Multiplicacao dos valores:  10.0


#### Questão 5

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e realize a divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.


In [13]:
valor1 = float(input(' Digite o primeiro valor: NUMERADOR '))
valor2 = float(input(' Digite o segundo valor: DENOMINADOR - NAO PODE SER ZERO !!! '))
print("Divisão dos valores: ", valor1 / valor2 )

 Digite o primeiro valor: NUMERADOR 12
 Digite o segundo valor: DENOMINADOR - NAO PODE SER ZERO !!! 4
Divisão dos valores:  3.0


#### Questão 6

Crie um programa que solicite dois valores numéricos, um operador e uma potência, e realize a exponenciação entre esses dois valores.


In [14]:
valor1 = float(input(' Digite o primeiro valor: OPERADOR '))
valor2 = float(input(' Digite o segundo valor: POTÊNCIA '))
print("Exponenciacao: ", valor1 ** valor2 )

 Digite o primeiro valor: OPERADOR 2
 Digite o segundo valor: POTÊNCIA 3
Exponenciacao:  8.0


#### Questão 7
Crie um programa que solicite dois valores numéricos, um numerador e um denominador e realize a divisão inteira entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [15]:
valor1 = float(input(' Digite o primeiro valor: NUMERADOR '))
valor2 = float(input(' Digite o segundo valor: DENOMINADOR - NAO PODE SER ZERO !!! '))
print("Divisão INTEIRA dos valores: ", valor1 // valor2 )

 Digite o primeiro valor: NUMERADOR 5
 Digite o segundo valor: DENOMINADOR - NAO PODE SER ZERO !!! 3
Divisão INTEIRA dos valores:  1.0


#### Questão 8

Crie um programa que solicite dois valores numéricos, um numerador e um denominador, e retorne o resto da divisão entre os dois valores. Deixe claro que o valor do denominador não pode ser `0`.

In [16]:
valor1 = float(input(' Digite o primeiro valor: NUMERADOR '))
valor2 = float(input(' Digite o segundo valor: DENOMINADOR - NAO PODE SER ZERO !!! '))
print("REsto da Divisão dos valores: ", valor1 % valor2 )

 Digite o primeiro valor: NUMERADOR 5
 Digite o segundo valor: DENOMINADOR - NAO PODE SER ZERO !!! 3
REsto da Divisão dos valores:  2.0


#### Questão 9

Crie um código que solicita 3 notas de um estudante e imprima a média das notas.

In [17]:
nota1 = float(input(' Digite a primeira nota'))
nota2 = float(input(' Digite a segunda nota'))
nota3 = float(input(' Digite a terceira nota'))
print("Média: ", (nota1+nota2+nota3)/3 )

 Digite a primeira nota8
 Digite a segunda nota9
 Digite a terceira nota9
Média:  8.666666666666666


#### Questão 10

Crie um código que calcule e imprima a média ponderada dos números 5, 12, 20 e 15 com pesos respectivamente iguais a 1, 2, 3 e 4.

In [18]:
media_ponderada = (5 * 1 + 12 * 2 + 20 * 3 + 15 * 4) / (1 + 2 + 3 + 4)
print("Média ponderada: ", media_ponderada)

Média ponderada:  14.9


### Editando textos

#### Questão 1

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase na tela.

In [19]:
frase = "we shall never surrender"
print(frase)

we shall never surrender


#### Questão 2
Crie um código que solicite uma frase e depois imprima a frase na tela.

In [20]:
frase2 = input("Digite uma frase: ")
print(frase2)

Digite uma frase: What a beautiful day
What a beautiful day


#### Questão 3

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras maiúsculas.

In [21]:
frase2 = input("Digite uma frase: ")
print(frase2.upper())

Digite uma frase: What a beautiful day
WHAT A BEAUTIFUL DAY


#### Questão 4

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase digitada mas com todas as letras minúsculas.

In [22]:
frase2 = input("Digite uma frase: ")
print(frase2.lower())

Digite uma frase: What a beautiful day
what a beautiful day


#### Questão 5

Crie uma variável chamada “`frase`” e atribua a ela uma string de sua escolha. Em seguida, imprima a frase sem espaços em branco no início e no fim.

In [23]:
frase = "     Vamos para Londres amanhã ?    "
print(frase.strip())

Vamos para Londres amanhã ?


#### Questão 6

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim.

In [24]:
frase2 = input("Digite uma frase: ")
print(frase2.strip())

Digite uma frase:            DEixei muitos espacos.   
DEixei muitos espacos.


#### Questão 7


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase sem espaços em branco no início e no fim e em letras minúsculas.

In [25]:
frase2 = input("Digite uma frase: ")
print(frase2.strip().lower())

Digite uma frase: PARE agora.   
pare agora.


#### Questão 8

Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`e`” trocadas pela letra “`f`”.

In [26]:
frase2 = input("Digite uma frase: ")
print(frase2.replace('e','f'))

Digite uma frase: Estádio de Futebol
Estádio df Futfbol


#### Questão 9


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as vogais “`a`” trocadas pela caractere  “`@`”.

In [27]:
frase2 = input("Digite uma frase: ")
print(frase2.replace('a','@'))

Digite uma frase: Cabana Aberta
C@b@n@ Abert@


#### Questão 10


Crie um código que solicite uma frase à pessoa usuária e imprima a mesma frase com todas as consoantes “`s`” trocadas pelo caractere  “`$`”.

In [28]:
frase2 = input("Digite uma frase: ")
print(frase2.replace('s','$'))

Digite uma frase: Sabor sensacional de peras. 
Sabor $en$acional de pera$. 


## Aula 3 - Estruturas condicionais

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça à pessoa usuária para fornecer dois números e exibir o número maior.

In [29]:

num1 = float(input('Digite o 1° número: '))

num2 = float(input('Digite o 2° número: '))

if num1 > num2:
  print(f'O número {num1} é maior que o número {num2}')
elif num1 < num2:
  print(f'O número {num2} é maior que o número {num1}')
else:
  print(f'Os números {num1} e {num2} são iguais')

Digite o 1° número: 3
Digite o 2° número: 8
O número 8.0 é maior que o número 3.0


#### Questão 2
Escreva um programa que solicite o percentual de crescimento de produção de uma empresa e informe se houve um crescimento (porcentagem positiva) ou decrescimento (porcentagem negativa).

In [30]:
crescimento = float(input('Digite o percentual de crescimento de produção da empresa: '))

if crescimento > 0:
  print(f'Houve um crescimento de {crescimento}%')

elif crescimento < 0:
  print(f'Houve um decrescimento de {crescimento}%')

else:
  print(f'Sem alteração no crescimento')

Digite o percentual de crescimento de produção da empresa: 12
Houve um crescimento de 12.0%


#### Questão 3
Escreva um programa que determine se uma letra fornecida pela pessoa usuária é uma vogal ou consoante.

In [31]:
vogais = 'a, e, i, o, u, A, E, I, O, U'

letra = input('Digite uma letra: ')

if letra in vogais:
  print(f'A letra {letra} é uma vogal')
else:
  print(f'A letra {letra} é uma consoante')

Digite uma letra: e
A letra e é uma vogal


#### Questão 4
Escreva um programa que leia valores médios de preços de um modelo de carro por 3 anos consecutivos e exiba o valor mais alto e mais baixo entre esses três anos.

In [32]:
ano1 = float(input('Digite o preço médio do carro no 1° ano: '))
ano2 = float(input('Digite o preço médio do carro no 2° ano: '))
ano3 = float(input('Digite o preço médio do carro no 3° ano: '))

if ano1 > ano2 and ano1 > ano3:
  print(f'O preço médio do carro no 1° ano é o mais alto: {ano1}')
elif ano2 > ano1 and ano2 > ano3:
  print(f'O preço médio do carro no 2° ano é o mais alto: {ano2}')
else:
   if ano3 > ano2 and ano3 > ano1:
      print(f'O preço médio do carro no 3° ano é o mais alto: {ano3}')

if ano1 < ano2 and ano1 < ano3:
  print(f'O preço médio do carro no 1° ano é o mais baixo: {ano1}')
elif ano2 < ano3 and ano2 < ano3:
  print(f'O preço médio do carro no 2 ano é o mais baixo: {ano2}')
else:
    print(f'O preço médio do carro no 3° ano é o mais baixo: {ano3}')

Digite o preço médio do carro no 1° ano: 120
Digite o preço médio do carro no 2° ano: 123
Digite o preço médio do carro no 3° ano: 144
O preço médio do carro no 3° ano é o mais alto: 144.0
O preço médio do carro no 1° ano é o mais baixo: 120.0


#### Questão 5
Escreva um programa que pergunte sobre o preço de três produtos e indique qual é o produto mais barato para comprar.

In [33]:
preco1 = float(input("Informe preço produto 1  "))
preco2 = float(input("Informe preço produto 2  "))
preco3 = float(input("Informe preço produto 3  "))

if preco1 < preco2 and preco1 < preco3:
  print(f'Produto 1 é o mais barato: {preco1}')

elif preco2 < preco1 and preco2 < preco3:
  print(f'Produto 2 é o mais barato: {preco2}')

else:
  print(f'Produto 3 é o mais barato: {preco3}')


Informe preço produto 1  23.90
Informe preço produto 2  44.00
Informe preço produto 3  45.00
Produto 1 é o mais barato: 23.9


#### Questão 6
Escreva um programa que leia três números e os exiba em ordem decrescente.

In [34]:
numero1 = float(input("Informe número 1  "))
numero2 = float(input("Informe número 2  "))
numero3 = float(input("Informe número 3  "))

if numero1 >= numero2 and numero1 >= numero3:
    dec1 = numero1
    if numero2 >= numero3:
        dec2 = numero2
        dec3 = numero3
    else:
        dec2 = numero3
        dec3 = numero2
elif numero2 >= numero1 and numero2 >= numero3:
    dec1 = numero2
    if numero1 >= numero3:
        dec2 = numero1
        dec3 = numero3
    else:
        dec2 = numero3
        dec3 = numero1
else:
    dec1 = numero3
    if numero1 >= numero2:
        dec2 = numero1
        dec3 = numero2
    else:
        dec2 = numero2
        dec3 = numero1

print(f'Números em ordem decrescente: {dec1}, {dec2}, {dec3}')


Informe número 1  87
Informe número 2  23
Informe número 3  45
Números em ordem decrescente: 87.0, 45.0, 23.0


#### Questão 7
Escreva um programa que pergunte em qual turno a pessoa usuária estuda ("manhã", "tarde" ou "noite") e exiba a mensagem "Bom Dia!", "Boa Tarde!", "Boa Noite!", ou "Valor Inválido!", conforme o caso.

In [35]:
turno = input('Em qual turno você estuda? ')

if turno == 'manhã':
  print('Bom dia!')
elif turno == 'tarde':
  print('Boa tarde!')
elif turno == 'noite':
  print('Boa noite!')
else:
  print('Valor inválido!')


Em qual turno você estuda? manhã
Bom dia!


#### Questão 8
Escreva um programa que peça um número inteiro à pessoa usuária e determine se ele é par ou ímpar. Dica: Você pode utilizar o operador módulo `%`.

In [36]:
numero = int(input('Digite um número inteiro: '))

if numero % 2 == 0:
  print(f'O número {numero} é par')
else:
  print(f'O número {numero} é ímpar')


Digite um número inteiro: 9
O número 9 é ímpar


#### Questão 9
Escreva um programa que peça um número à pessoa usuária e informe se ele é inteiro ou decimal.

In [37]:
numero = float(input('Digite um número: '))

if numero == int(numero):
  print(f'O número {numero} é inteiro')
else:
  print(f'O número {numero} é decimal')


Digite um número: 9.867
O número 9.867 é decimal


### Momento dos projetos

#### Questão 10
Um programa deve ser escrito para ler dois números e, em seguida, perguntar à pessoa usuária qual operação ele deseja realizar. O resultado da operação deve incluir informações sobre o número - se é par ou ímpar, positivo ou negativo e inteiro ou decimal.

In [39]:
numero1 = float(input('Digite o 1° número: '))
numero2 = float(input('Digite o 2° número: '))

operacao = input('Qual operação deseja realizar? ')

if operacao == '+':
  resultado = numero1 + numero2
elif operacao == '-':
  resultado = numero1 - numero2
elif operacao == '*':
  resultado = numero1 * numero2
elif operacao == '/':
  resultado = numero1 / numero2

if resultado % 2 == 0:
  print(f'O resultado da operação é {resultado} e é par')
else:
  print(f'O resultado da operação é {resultado} e é ímpar')

if resultado > 0:
  print(f'O resultado da operação é {resultado} e é positivo')
else:
  print(f'O resultado da operação é {resultado} e é negativo')

if resultado == int(resultado):
  print(f'O resultado da operação é {resultado} e é inteiro')
else:
  print(f'O resultado da operação é {resultado} e é decimal')



Digite o 1° número: 9
Digite o 2° número: 8
Qual operação deseja realizar? +
O resultado da operação é 17.0 e é ímpar
O resultado da operação é 17.0 e é positivo
O resultado da operação é 17.0 e é inteiro


#### Questão 11
Escreva um programa que peça à pessoa usuária três números que representam os lados de um triângulo. O programa deve informar se os valores podem ser utilizados para formar um triângulo e, caso afirmativo, se ele é equilátero, isósceles ou escaleno. Tenha em mente algumas dicas:
- Três lados formam um triângulo quando a soma de quaisquer dois lados for maior que o terceiro;
- Triângulo Equilátero: três lados iguais;
- Triângulo Isósceles: quaisquer dois lados iguais;
- Triângulo Escaleno: três lados diferentes;


In [40]:

lado1 = float(input('Digite o 1° lado do triângulo: '))
lado2 = float(input('Digite o 2° lado do triângulo: '))
lado3 = float(input('Digite o 3° lado do triângulo: '))

if lado1 + lado2 > lado3 or lado1 + lado3 > lado2 or lado2 + lado3 > lado1:
  print('É possível formar um triângulo')

if lado1 == lado2 == lado3:
    print('É um triângulo equilátero')
elif lado1 == lado2 or lado1 == lado3 or lado2 == lado3:
    print('É um triângulo isósceles')
elif lado1 != lado2 != lado3:
    print('É um triângulo escaleno')


Digite o 1° lado do triângulo: 12
Digite o 2° lado do triângulo: 3
Digite o 3° lado do triângulo: 4
É possível formar um triângulo
É um triângulo escaleno


#### Questão 12
Um estabelecimento está vendendo combustíveis com descontos variados. Para o etanol, se a quantidade comprada for até 15 litros, o desconto será de 2% por litro. Caso contrário, será de 4% por litro. Para o diesel, se a quantidade comprada for até 15 litros, o desconto será de 3% por litro. Caso contrário, será de 5% por litro. O preço do litro de diesel é R\$ 2,00 e o preço do litro de etanol é R\$ 1,70. Escreva um programa que leia a quantidade de litros vendidos e o tipo de combustível (E para etanol e D para diesel) e calcule o valor a ser pago pelo cliente. Tenha em mente algumas dicas:

- O do valor do desconto será a multiplicação entre preço do litro, quantidade de litros e o valor do desconto.
- O valor a ser pago por um cliente será o resultado da multiplicação do preço do litro pela quantidade de litros menos o valor de desconto resultante do cálculo.


In [41]:

litros = 0.00
litros = float(input('Digite a quantidade de litros: '))

tipo = input('Digite o tipo de combustível (E para etanol e D para diesel): ')

if tipo == 'E':
  if litros <= 15:
    desconto = 0.02
  else:
    desconto = 0.04
  preco = 1.70
  pagar = (litros * preco) - (litros * preco * desconto)
  print(f'O valor a ser pago é R$ {pagar:.2f}')
elif tipo == 'D':
  if litros <= 15:
    desconto = 0.03
  else:
    desconto = 0.05
    preco = 2.00
  pagar = (litros * preco) - (litros * preco * desconto)
  print(f'O valor a ser pago é R$ {pagar:.2f}')
else:
  print('Tipo de combustível inválido')

Digite a quantidade de litros: 55
Digite o tipo de combustível (E para etanol e D para diesel): D
O valor a ser pago é R$ 104.50


#### Questão 13
Em uma empresa de venda de imóveis você precisa criar um código que analise os dados de vendas anuais para ajudar a diretoria na tomada de decisão. O código precisa coletar os dados de quantidade de venda durante os anos de 2022 e 2023 e fazer um cálculo de variação percentual. A partir do valor da variação, deve ser enviada às seguintes sugestões:
- Para variação acima de 20%: bonificação para o time de vendas.
- Para variação entre 2% e 20%: pequena bonificação para time de vendas.
- Para variação entre 2% e -10%: planejamento de políticas de incentivo às vendas.
- Para bonificações abaixo de -10%: corte de gastos.


In [42]:
quantidade_vendas_2022 = float(input('Digite a quantidade de vendas em 2022: '))

quantidade_vendas_2023 = float(input('Digite a quantidade de vendas em 2023: '))

variacao = ((quantidade_vendas_2023 - quantidade_vendas_2022) / quantidade_vendas_2022) * 100

if variacao > 20:
   print('Bonificação para o time de vendas')
elif variacao >= 2 and variacao <= 20:
   print('Pequena bonificação para time de vendas')
elif variacao >= -10 and variacao < 2:
   print('Planejamento de políticas de incentivo às vendas')
else:
    print('Corte de gastos')


Digite a quantidade de vendas em 2022: 125000000
Digite a quantidade de vendas em 2023: 120000000
Planejamento de políticas de incentivo às vendas


## Aula 4 - Estruturas de repetição

### Aquecendo na programação

#### Questão 1
Escreva um programa que peça dois números inteiros e imprima todos os números inteiros entre eles.

In [43]:

num1 = int(input("Digite o 1° número: "))
num2 = int(input("Digite o 2° número: "))

for i in range(num1, (num2 + 1)):
  print(i)


Digite o 1° número: 8
Digite o 2° número: 12
8
9
10
11
12


#### Questão 2
Escreva um programa para calcular quantos dias levará para a colônia de uma bactéria A ultrapassar ou igualar a colônia de uma bactéria B, com base nas taxas de crescimento de 3% e 1,5% respectivamente. Considere que a colônia A inicia com 4 elementos e a colônia B com 10 elementos.

In [44]:

colonia_a = 4
colonia_b = 10

taxa_a = 0.03
taxa_b = 0.015

dias = 0

while colonia_a <= colonia_b:
  colonia_a *= 1 + taxa_a
  colonia_b *= 1 + taxa_b
  dias += 1

print(f'Serão necessários {dias} dias para a colônia A ultrapassar a colônia B')


Serão necessários 63 dias para a colônia A ultrapassar a colônia B


#### Questão 3
Para tratar uma quantidade de 15 dados de avaliações de pessoas usuárias de um serviço da empresa, precisamos verificar se as notas são válidas. Então, escreva um programa que vai receber a nota de 0 a 5 de todos os dados e verificar se é um valor válido. Caso seja inserido uma nota acima de 5 ou abaixo de 0, repita até que a pessoa usuária insira um valor válido.

In [45]:
nota = 0.0

# for cont in range(16):
for cont in range(6):
  nota = float(input(f"Digite nota usuario {cont + 1}: "))

  while nota < 0 or nota > 5:
    nota = float(input(f"Nota inválida. Digite novamente para user {cont + 1}: "))

print('Verificação feita. Todas as notas são válidas')



Digite nota usuario 1: 4
Digite nota usuario 2: 6
Nota inválida. Digite novamente para user 2: 7
Nota inválida. Digite novamente para user 2: 7
Nota inválida. Digite novamente para user 2: 2
Digite nota usuario 3: 2
Digite nota usuario 4: 2
Digite nota usuario 5: 3
Digite nota usuario 6: 2
Verificação feita. Todas as notas são válidas


#### Questão 4
Desenvolva um programa que leia um conjunto indeterminado de temperaturas em Celsius e faça uma análise. Portanto, escreva um programa que leia temperaturas e informe a média delas. A leitura deve ser encerrada ao ser enviado o valor -273°C.

In [46]:

temperatura = float(input("Digite uma temperatura em Celsius: "))

soma = 0
cont = 0

while temperatura != -273:
  soma += temperatura
  cont += 1
  temperatura = float(input("Digite uma temperatura em Celsius: "))

media = soma / cont

print(f'A média das temperaturas é {media:.2f}')


Digite uma temperatura em Celsius: 34
Digite uma temperatura em Celsius: 37
Digite uma temperatura em Celsius: 45
Digite uma temperatura em Celsius: -273
A média das temperaturas é 38.67


#### Questão 5
Escreva um programa que calcule o fatorial de um número inteiro fornecido pela pessoa usuária. O fatorial de um número inteiro é a multiplicação desse número por todos os seus antecessores até o número 1. Por exemplo, o fatorial de 5 é 5 x 4 x 3 x 2 x 1 = 120.

In [49]:
num = int(input("Digite um número inteiro: "))

fatorial = 1

i = num

while i > 0:
  fatorial *= i
  i -= 1

print(f'O fatorial de {num} é {fatorial}')



Digite um número inteiro: 9
O fatorial de 9 é 362880


### Momento dos projetos

#### Questão 6
Escreva um programa que gere a tabuada de um número inteiro de 1 a 10, de acordo com a escolha da pessoa usuária. Como exemplo, para o número 2, a tabuada deve ser mostrada no seguinte formato:
```
Tabuada do 2:
2 x 1 = 2
2 x 2 = 4
[...]
2 x 10 = 20
```

In [48]:
num = int(input('Informe um número inteiro de 1 a 10: '))

print(f'Tabuada do número {num}:')

for i in range(1, 11):
    result = num * i
    print(f'{num} x {i} = {result}')


Informe um número inteiro de 1 a 10: 5
Tabuada do número 5:
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50


#### Questão 7
Os números primos possuem várias aplicações dentro da Ciência de Dados, por exemplo, na criptografia e segurança. Um número primo é aquele que é divisível apenas por um e por ele mesmo. Faça um programa que peça um número inteiro e determine se ele é ou não um número primo.

In [50]:
# Solicitamos o número
num = int(input("Digite o número: "))

# Assumimos que o número é primo até que se prove o contrário
eh_primo = True

# Números abaixo de 2 não são primos
if num <= 1 :
  eh_primo = False
else:
    for i in range(2, num):
        # Se o número for divisível por qualquer número dentro deste intervalo,
        # ele não é primo, portanto, mudamos a variável 'eh_primo' para False e saímos do loop.
        if num % i == 0:
            eh_primo = False
            break

# Verifica se 'eh_primo' ainda é True, o que significa que o número passou pelo loop
# sem ser divisível por nenhum número além de 1 e ele mesmo.
if eh_primo:
    # Se for o caso, o número é primo.
    print(f'O número {num} é primo')
else:
    # Caso contrário, o número não é primo.
    print(f'O número {num} não é primo')


Digite o número: 11
O número 11 é primo


#### Questão 8
Vamos entender a distribuição de idades de pensionistas de uma empresa de previdência. Escreva um programa que leia as idades de uma quantidade não informada de clientes e mostre a distribuição em intervalos de [0-25], [26-50], [51-75] e [76-100]. Encerre a entrada de dados com um número negativo.

In [51]:

idade = int(input('Informe a idade ou número negativo para finalizar:'))

contador_0_25 = 0
contador_26_50 = 0
contador_51_75 = 0
contador_76_100 = 0

while idade >= 0:
    if idade >= 0 and idade <= 25:
        contador_0_25 += 1
    elif idade >= 26 and idade <= 50:
        contador_26_50 += 1
    elif idade >= 51 and idade <= 75:
        contador_51_75 += 1
    elif idade >= 76 and idade <= 100:
        contador_76_100 += 1


    idade = int(input('Informe a idade ou número negativo para finalizar:'))

print('Distribuição de idades:')
print('0-25:', contador_0_25)
print('26-50:', contador_26_50)
print('51-75:', contador_51_75)
print('76-100:', contador_76_100)

Informe a idade ou número negativo para finalizar:88
Informe a idade ou número negativo para finalizar:98
Informe a idade ou número negativo para finalizar:-1
Distribuição de idades:
0-25: 0
26-50: 0
51-75: 0
76-100: 2


#### Questão 9
Em uma eleição para gerência em uma empresa com 20 funcionários, existem quatro candidatos. Escreva um programa que calcule o vencedor da eleição. A votação ocorreu da seguinte maneira:
- Cada funcionário votou em um dos quatro candidatos (representados pelos números 1, 2, 3 e 4).
- Também foram contabilizados os votos nulos (representado pelo número 5) e os votos em branco (representado pelo número 6).

Ao final da votação, o programa deve exibir o total de votos para cada candidato, o número de votos nulos e o número de votos em branco. Além disso, deve calcular e exibir a porcentagem de votos nulos em relação ao total de votos e a porcentagem de votos em branco em relação ao total de votos.

In [53]:

votos_candidato1 = 0
votos_candidato2 = 0
votos_candidato3 = 0
votos_candidato4 = 0

votos_nulos = 0
votos_branco = 0

for i in range(0,20):
    voto = int(input('Informe voto: '))
    if voto == 1:
        votos_candidato1 += 1
    elif voto == 2:
        votos_candidato2 += 1
    elif voto == 3:
        votos_candidato3 += 1
    elif voto == 4:
        votos_candidato4 += 1
    elif voto == 5:
        votos_nulos += 1
    elif voto == 6:
        votos_branco += 1
    else:
        print("Voto inválido.")

print(f'Votos candidato 1: {votos_candidato1}')
print(f'Votos candidato 2: {votos_candidato2}')
print(f'Votos candidato 3: {votos_candidato3}')
print(f'Votos candidato 4: {votos_candidato4}')
print(f'Votos nulos: {votos_nulos}')
print(f'Votos em branco: {votos_branco}')
print(f'Percentual de votos nulos: {(votos_nulos / 20 * 100)}')
print(f'Percentual de votos em branco: {(votos_branco / 20 * 100)}')

Informe voto: 1
Informe voto: 2
Informe voto: 3
Informe voto: 4
Informe voto: 3
Informe voto: 2
Informe voto: 1
Informe voto: 2
Informe voto: 3
Informe voto: 4
Informe voto: 3
Informe voto: 3
Informe voto: 4
Informe voto: 3
Informe voto: 4
Informe voto: 2
Informe voto: 2
Informe voto: 1
Informe voto: 2
Informe voto: 2
Votos candidato 1: 3
Votos candidato 2: 7
Votos candidato 3: 6
Votos candidato 4: 4
Votos nulos: 0
Votos em branco: 0
Percentual de votos nulos: 0.0
Percentual de votos em branco: 0.0


## Aula 5 - Estruturas de dados

### Aquecendo na programação

#### Questão 1
Faça um programa que tenha a seguinte lista contendo os valores de gastos de uma empresa de papel `[2172.54, 3701.35, 3518.09, 3456.61, 3249.38, 2840.82, 3891.45, 3075.26, 2317.64, 3219.08]`. Com esses valores, faça um programa que calcule a média de gastos. Dica, use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).

#### Questão 2
 Com os mesmos dados da questão anterior, defina quantas compras foram acima de 3000 reais e calcule a porcentagem quanto ao total de compras.

#### Questão 3
Faça um código que colete em uma lista 5 números inteiros quaisquer e imprima a lista. Exemplo: `[1,4,7,2,4]`.

#### Questão 4
Colete novamente 5 inteiros e imprima a lista em ordem inversa à enviada.

#### Questão 5
Faça um programa que, ao inserir um número qualquer, criará uma lista contendo todos os números primos entre 1 e o número digitado.

#### Questão 6
Escreva um programa que peça uma data informando o dia, mês e ano e determine se ela é uma data válida para uma análise.

### Momento dos projetos

#### Questão 7
Para um estudo envolvendo o nível de multiplicação de bactérias em uma colônia, foi coletado o número de bactérias multiplicadas por dia, com base em um processo de duplicação diária e pode ser observado a seguir: `[1.2, 2.1, 3.3, 5.0, 7.8, 11.3, 16.6, 25.1, 37.8, 56.9]`. Tendo esses valores, faça um código que gere uma lista contendo o percentual de crescimento de bactérias por dia, comparando o número de bactérias em cada dia com o número de bactérias do dia anterior. Dica: para calcular o percentual de crescimento usamos a seguinte equação: `100 * (amostra_atual - amostra_passada) / (amostra_passada)`.

#### Questão 8
Para uma seleção de produtos alimentícios, precisamos separar o conjunto de IDs dados por números inteiros, sabendo que os produtos com ID par são doces e os com ID ímpar são amargos. Monte um código que colete 10 IDs. Depois, calcule e mostre a quantidade de produtos doces e amargos.

#### Questão 9
Desenvolva um programa que informa a nota de um aluno de acordo com suas respostas. Ele deve pedir a resposta de um aluno para cada questão e é preciso verificar se a resposta foi igual ao gabarito. Cada questão vale um ponto e existem alternativas:  A, B, C ou D.
```
Gabarito da Prova:
01 - D
02 - A
03 - C
04 - B
05 - A
06 - D
07 - C
08 - C
09 - A
10 - B
```

#### Questão 10
Um instituto de meteorologia deseja fazer um estudo de temperatura média de cada mês do ano. Para isso, você precisa fazer um código que colete e armazene essas temperaturas médias em uma lista. Depois, calcule a média anual das temperaturas e mostre todas as temperaturas acima da média anual, e em que mês elas ocorreram, mostrando os meses por extenso: Janeiro, Fevereiro, etc.

#### Questão 11
Uma empresa de e-commerce está interessada em analisar as vendas dos seus produtos. Os dados das vendas foram armazenados em um dicionário:
```
{'Produto A': 300, 'Produto B': 80, 'Produto C': 60,
 'Produto D': 200, 'Produto E': 250, 'Produto F': 30}
```
Escreva um código que calcule o total de vendas e o produto mais vendido.

#### Questão 12
Uma pesquisa de mercado foi feita para decidir qual design de uma marca infantil mais agrada crianças. A pesquisa foi feita e o votos computados podem ser observados abaixo:
```
'''
Tabela de votos da marca
Design 1 - 1334 votos
Design 2 - 982 votos
Design 3 - 1751 votos
Design 4 - 210 votos
Design 5 - 1811 votos
'''
```
Adapte os dados fornecidos a você para uma estrutura de dicionário e a partir dele, informe o design vencedor e a porcentagem de votos recebidos.


#### Questão 13
Os funcionários de um setor da empresa que você trabalha vão receber um abono correspondente a 10% do seu salário devido ao ótimo desempenho do time. O setor financeiro solicitou sua ajuda para a verificação das consequências financeiras que esse abono irá gerar nos recursos. Assim, foi encaminhada para você uma lista com os salários que receberão o abono: `[1172, 1644, 2617, 5130, 5532, 6341, 6650, 7238, 7685, 7782, 7903]`. O abono de cada funcionário não pode ser inferior a 200. Em código, transforme cada um dos salários em chaves de um dicionário e o abono de cada salário no elemento. Depois, informe o total de gastos com o abono, quantos funcionários receberam o abono mínimo e qual o maior valor de abono fornecido.


#### Questão 14
Uma equipe de cientistas de dados está estudando a diversidade biológica em uma floresta. A equipe fez a coleta de informações sobre o número de espécies de plantas e animais em cada área da floresta e armazenaram essas informações em um dicionário. Nele, a chave descreve a área dos dados e os valores nas listas correspondem às espécies de plantas e animais nas áreas, respectivamente.
```
{'Área Norte': [2819, 7236],
 'Área Leste': [1440, 9492],
 'Área Sul': [5969, 7496],
 'Área Oeste': [14446, 49688],
 'Área Centro': [22558, 45148]}
```
Escreva um código para calcular a média de espécies por área e identificar a área com a maior diversidade biológica. Dica: use as funções built-in [`sum()`](https://docs.python.org/3/library/functions.html#sum) e [`len()`](https://docs.python.org/3/library/functions.html#len).


#### Questão 15
O setor de RH da sua empresa te pediu uma ajuda para analisar as idades dos funcionários de 4 setores da empresa. Para isso, ele te forneceu os seguintes dados:
```
{'Setor A': [22, 26, 30, 30, 35, 38, 40, 56, 57, 65],
 'Setor B': [22, 24, 26, 33, 41, 49, 50, 54, 60, 64],
 'Setor C': [23, 26, 26, 29, 34, 35, 36, 41, 52, 56],
 'Setor D': [19, 20, 25, 27, 34, 39, 42, 44, 50, 65]}
```
Sabendo que cada setor tem 10 funcionários, construa um código que calcule a média de idade de cada setor, a idade média geral entre todos os setores e quantas pessoas estão acima da idade média geral.
